staking/slash/double-sign.go

```diff
@@ -1,52 +1,291 @@
+	Reporter common.Address `json:"reporter"`
+	Offender common.Address `json:"offender"`
+}
+
+// Application ..
+type Application struct {
+	TotalSlashed, TotalSnitchReward *big.Int
+}
+
+func (a *Application) String() string {
+	s, _ := json.Marshal(a)
+	return string(s)
+}
+
+// MarshalJSON ..
+func (e Evidence) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Moment
+		ProposalHeader string `json:"header"`
+	}{e.Moment, e.ProposalHeader.String()})
+}
+
+// Records ..
+type Records []Record
+
+func (r Records) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
+
+// MarshalJSON ..
+func (r Record) MarshalJSON() ([]byte, error) {
+	reporter, offender :=
+		common2.MustAddressToBech32(r.Reporter),
+		common2.MustAddressToBech32(r.Offender)
+	return json.Marshal(struct {
+		ConflictingBallots
+		Evidence         Evidence `json:"evidence"`
+		Beneficiary      string   `json:"beneficiary"`
+		AddressForBLSKey string   `json:"offender"`
+	}{r.ConflictingBallots, r.Evidence, reporter, offender})
+}
+
+func (e Evidence) String() string {
+	s, _ := json.Marshal(e)
+	return string(s)
+}
+
+func (r Record) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
 
 // Verify checks that the signature is valid
 func Verify(candidate *Record) error {
+
+	fmt.Println("need to verify the slash", candidate)
+
 	return nil
 }
 
-// Apply ..
-func Apply(state *state.DB, slashes []byte) error {
+var (
+	oneHundredPercent = numeric.NewDec(1)
+	fiftyPercent      = numeric.MustNewDecFromStr("0.5")
+)
+
+// applySlashRate returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func applySlashRate(
+	amount *big.Int, rate numeric.Dec,
+) (*big.Int, *big.Int, *big.Int) {
+	amountPostSlash := numeric.NewDecFromBigInt(
+		amount,
+	).Mul(numeric.OneDec().Sub(rate)).TruncateInt()
+	amountOfSlash := new(big.Int).Sub(amount, amountPostSlash)
+	return amountPostSlash, amountOfSlash, new(big.Int).Div(amountOfSlash, common.Big2)
+}
+
+// postSlashBalanceCurrentDelegator applySlashRate
+// returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func postSlashBalanceCurrentDelegator(
+	state *state.DB,
+	validatorAddr common.Address,
+	delegatorAddr common.Address,
+	rate numeric.Dec,
+) (*big.Int, *big.Int, *big.Int) {
+	if wrapper := state.GetStakingInfo(validatorAddr); wrapper != nil {
+		for _, delegation := range wrapper.Delegations {
+			if delegation.DelegatorAddress == delegatorAddr {
+				return applySlashRate(delegation.Amount, rate)
+			}
+		}
+	}
+	fmt.Println("invariant violated, should not come here")
+	return nil, nil, nil
+}
+
+func delegatorSlashApply(
+	delegatedToValidator *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	doubleSignEpoch *big.Int,
+	slashTrack *Application,
+) error {
+	for _, delegationSnapshot := range delegatedToValidator.Delegations {
+		postSlashBal, slashDebt, halfOfSlashDebt := postSlashBalanceCurrentDelegator(
```

Comment:
This is not correct. when you calculate how much to slash, you should use the snapshot. Not the current state.

Reply:
And here, since you are using the current state, and applying the rate on it, postSlashBal will always be positive.

Reply:
fixed